File: common/src/main/java/guepardoapps/lucahome/common/converter/JsonDataToRoomConverter.java
package guepardoapps.lucahome.common.converter;

import android.support.annotation.NonNull;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.UUID;

import guepardoapps.lucahome.common.classes.Room;
import guepardoapps.lucahome.common.classes.ILucaClass;
import guepardoapps.lucahome.common.utils.Logger;
import guepardoapps.lucahome.common.utils.StringHelper;

public class JsonDataToRoomConverter implements IJsonDataConverter {
    private static final String Tag = JsonDataToRoomConverter.class.getSimpleName();
    private static final String SearchParameter = "{\"Data\":";

    private static final JsonDataToRoomConverter Singleton = new JsonDataToRoomConverter();

    public static JsonDataToRoomConverter getInstance() {
        return Singleton;
    }

    private JsonDataToRoomConverter() {
    }

    @Override
    public ArrayList<Room> GetList(@NonNull String[] jsonStringArray) {
        if (StringHelper.StringsAreEqual(jsonStringArray)) {
            return parseStringToList(jsonStringArray[0]);
        } else {
            String usedEntry = StringHelper.SelectString(jsonStringArray, SearchParameter);
            return parseStringToList(usedEntry);
        }
    }

    @Override
    public ArrayList<Room> GetList(@NonNull String responseString) {
        return parseStringToList(responseString);
    }

    private ArrayList<Room> parseStringToList(@NonNull String jsonString) {
        if (!jsonString.contains("Error")) {
            ArrayList<Room> list = new ArrayList<>();

            try {
                JSONObject jsonObject = new JSONObject(jsonString);
                JSONArray dataArray = jsonObject.getJSONArray("Data");

                for (int index = 0; index < dataArray.length(); index++) {
                    JSONObject child = dataArray.getJSONObject(index).getJSONObject("Room");

                    UUID uuid = UUID.fromString(child.getString("Uuid"));

                    String name = child.getString("Name");
                    int roomTypeInteger = child.getInt("RoomType");

                    Room.RoomType roomType = Room.RoomType.values()[roomTypeInteger];

                    Room newRoom = new Room(uuid, name, roomType, true, ILucaClass.LucaServerDbAction.Null);
                    list.add(newRoom);
                }
            } catch (JSONException jsonException) {
                Logger.getInstance().Error(Tag, jsonException.getMessage());
            }

            return list;
        }

        Logger.getInstance().Error(Tag, jsonString + " has an error!");
        return new ArrayList<>();
    }
}

File: common/src/main/java/guepardoapps/lucahome/common/converter/room/JsonDataToRoomConverter.kt
package guepardoapps.lucahome.common.converter.room

import guepardoapps.lucahome.common.annotations.JsonKey
import guepardoapps.lucahome.common.enums.room.RoomType
import guepardoapps.lucahome.common.extensions.room.getJsonKey
import guepardoapps.lucahome.common.extensions.room.getPropertyJsonKey
import guepardoapps.lucahome.common.models.room.Room
import guepardoapps.lucahome.common.utils.Logger
import org.json.JSONArray
import org.json.JSONObject
import java.util.*

internal class JsonDataToRoomConverter {
    private val tag = JsonDataToRoomConverter::class.java.simpleName

    fun parse(jsonResponse: String): ArrayList<Room> {
        val list: ArrayList<Room> = ArrayList()

        try {
            val jsonObject = JSONObject(jsonResponse)

            val room = Room()
            val jsonKey: JsonKey = room.getJsonKey()
            val dataArray: JSONArray = jsonObject.getJSONArray(jsonKey.parent)

            for (index: Int in 0..dataArray.length()) {
                val value: JSONObject = dataArray.getJSONObject(index)
                val data: JSONObject = value.getJSONObject(Room::class.java.simpleName)

                val uuidJsonKey = room.getPropertyJsonKey(room::uuid.name)
                val nameJsonKey = room.getPropertyJsonKey(room::name.name)
                val typeJsonKey = room.getPropertyJsonKey(room::type.name)

                room.uuid = UUID.fromString(data.getString(uuidJsonKey.key))
                room.name = data.getString(nameJsonKey.key)
                room.type = RoomType.values()[data.getInt(typeJsonKey.key)]

                list.add(room)
            }
        } catch (exception: Exception) {
            Logger.instance.error(tag, exception)
        } finally {
            return list
        }
    }
}